### PR TITLE
Add couchdbreceiver factory and config

### PR DESCRIPTION
**Pull request by cpheps**
**Description:** <Describe what has changed.>

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -639,6 +639,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/journ
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zookeeperreceiver => ./receiver/zookeeperreceiver
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/apachereceiver => ./receiver/apachereceiver
```
Comment (cpheps) [inline go.mod:642]: This is unrelated to couchdb likely needs to be removed.

File: receiver/couchdbreceiver/config.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	"net/url"
+	"strings"
+
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+	"go.uber.org/multierr"
+)
+
+var (
+	defaultEndpoint = "http://localhost:5984"
```
Comment (cpheps) [inline receiver/couchdbreceiver/config.go:29]: This should probably be a `const` unless you plan on overriding it for tests.

File: receiver/couchdbreceiver/config.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	if u.Hostname() == "" {
+		return fmt.Errorf(errInvalidHost.Error(), cfg.Endpoint)
+	}
+
+	return nil
+}
+
+// validSchema returns true if any http protocol is found.
+func validSchema(rawUrl string) bool {
+	lowerUrl := strings.ToLower(rawUrl)
```
Comment (cpheps) [inline receiver/couchdbreceiver/config.go:93]: I'm note sure what the Otel way of parsing a URL is but you could use the [url](https://pkg.go.dev/net/url#URL) package of Go stdlib to check all of this.

**Review by cpheps — COMMENTED** [review]
Overall looks good just a few nitpick things. Might be work asking @djaglowski for a quick review before we send it along.

File: receiver/couchdbreceiver/Makefile [hunk 4]
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
Comment (djaglowski) [inline receiver/couchdbreceiver/Makefile:1]: You'll always need to include a newline

File: receiver/couchdbreceiver/config.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	return strings.HasPrefix(lowerUrl, "http://") || strings.HasPrefix(lowerUrl, "https://")
+}
```
Comment (djaglowski) [inline receiver/couchdbreceiver/config.go:95]: It's not clear why validation is so complicated in this receiver vs others. Probably can be simplified quite a bit, depend on standard validation, etc.

File: receiver/couchdbreceiver/factory.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+			},
+			Endpoint: defaultEndpoint,
+			Timeout:  10 * time.Second,
+		},
+		AllNodes: false,
+	}
+}
+
+func createMetricsReceiver(_ context.Context, params component.ReceiverCreateSettings, rConf config.Receiver, consumer consumer.Metrics) (component.MetricsReceiver, error) {
+	return nil, nil // TODO build and return receiver in followup PR
```
Comment (djaglowski) [inline receiver/couchdbreceiver/factory.go:62]: TODOs are not allowed, even in cases like this. Just remove it.

File: receiver/couchdbreceiver/go.mod [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+require (
+	go.opentelemetry.io/collector v0.40.0
+	go.uber.org/multierr v1.7.0
+)
+
+require github.com/stretchr/testify v1.7.0
+
+require go.opentelemetry.io/collector/model v0.40.0
+
+require (
```
Comment (djaglowski) [inline receiver/couchdbreceiver/go.mod:14]: Run `go mod tidy` to ensure this is correct

File: receiver/couchdbreceiver/metadata.yaml [hunk 8]
```diff
@@ -0,0 +1 @@
+name: couchdb
```
Comment (djaglowski) [inline receiver/couchdbreceiver/metadata.yaml:1]:
```suggestion
name: couchdb

```

File: go.mod (hunk 1 above)
Comment (JonathanWamsley) [reply go.mod]: will remove

File: receiver/couchdbreceiver/config.go (hunk 2 above)
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/config.go]: will update to const

File: receiver/couchdbreceiver/config.go (hunk 3 above)
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/config.go]: I saw url.Parse def say it may not return errors for invalid schema's so I handled that with a validSchema function.  I'll remove though. Screen Shot 2021-12-13 at 9 54 03 AM

File: receiver/couchdbreceiver/Makefile (hunk 4 above)
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/Makefile]: will add

File: receiver/couchdbreceiver/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+> :construction: This receiver is in **BETA**. Configuration fields and metric data model are subject to change.
+
+## Prerequisites
+
+This receiver supports Couchdb versions `2.3+` and `3.1+`.
+
+## Configuration
+
+The following settings are required:
+- `endpoint` (default: `http://localhost:5984`): The URL of the couchdb endpoint
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/README.md]: will move to being optional

File: receiver/couchdbreceiver/config.go (hunk 5 above)
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/config.go]: I'll move everything into one function to make it look simplified. The username and password check is reused from other receivers like postgres. The url.Parse and hostname check is similar to apache. The validSchema is new and will be removed.

File: receiver/couchdbreceiver/factory.go (hunk 6 above)
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/factory.go]: will remove

File: receiver/couchdbreceiver/metadata.yaml (hunk 8 above)
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/metadata.yaml]: added new line. I believe this should be couchdbreceiver right?

File: receiver/couchdbreceiver/testdata/config.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+exporters:
+  nop:
+
+service:
+  pipelines:
+    metrics:
+     receivers: [couchdb]
+     processors: [nop]
+     exporters: [nop]
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply receiver/couchdbreceiver/testdata/config.yaml]: added new line

File: versions.yaml [hunk 11]
```diff
@@ -81,6 +81,7 @@ module-sets:
       - github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver
       - github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver
       - github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver
+      - github.com/open-telemetry/opentelemetry-collector-contrib/receiver/couchdbreceiver
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply versions.yaml]: remove it for now then? done

File: receiver/couchdbreceiver/config.go (hunk 3 above)
Comment (cpheps) [reply receiver/couchdbreceiver/config.go]: That's fair, there are a lot of schemas might be hard to determine an "invalid one". Since we only care about http or https should be fine.
